04-onion-routing.md

```diff
@@ -76,81 +70,59 @@ The use of a fixed nonce is safe since the keys are never reused.
 
 ~~~~
-+--------+-----------------+
-| header | per-hop payload |
-+--------+-----------------+
++------------------+-----------------------+------------------------+-----------------+
+| version (1 byte) | public key (33 bytes) | hops_data (20x53 bytes) | HMAC (20 bytes) |
++------------------+-----------------------+------------------------+-----------------+
 ~~~~
 
-The header is a fixed 854 byte array containing the necessary information for each hop to identify the next hop, and verify the integrity of the packet.
-It consists of a version byte, a 33 byte compressed `secp256k1` public key, used during the shared secret generation, a 20 byte HMAC used to verify the packet's integrity and an 800 byte routing information field.
 For this specification the version byte has a constant value of `0x00`.
 
-~~~~
-+------------------+-----------------------+-----------------+------------...-----------+
-| Version (1 byte) | Public Key (33 bytes) | HMAC (20 bytes) | Routing Info (800 bytes) |
-+------------------+-----------------------+-----------------+------------...-----------+
-~~~~
-
-The routing info field is a structure that holds obfuscated versions of the next hop's address and the associated HMAC.
-It is 800 bytes long, i.e., 20 byte MAC and 20 byte address times 20 hops, and has the following structure:
+The `hops_data` field is a structure that holds obfuscated versions of the next hop's address, transfer information and the associated HMAC.  It is 1060 bytes long, and has the following structure:
 
 ~~~~
-+-------------+----------+-------------+----------+-----+--------+
-| n_1 address | n_1 HMAC | n_2 address | n_2 HMAC | ... | filler |
-+-------------+----------+-------------+----------+-----+--------+
++-----------+-------------+----------+-----------+-------------+----------+-----------+
+| n_1 realm | n_1 per-hop | n_1 HMAC | n_2 realm | n_2 per-hop | n_2 HMAC | ...filler |
++-----------+-------------+----------+----------+--------------+----------+------------+
 ~~~~
 
-Where the `filler` consists of obfuscated deterministically generated
+Where `per-hop` is 32 bytes whose contents depend on `realm`, and `filler` consists of obfuscated deterministically generated
 padding. For details about how the `filler` is generated please see
 below. In addition, every _(address, HMAC)_-pair is incrementally
 obfuscated at each hop.
 
-The per-hop payloads has a similar structure:
+The `realm` byte determines the format of the `per-hop`; so far
+only `realm` 0 is defined, and for that, the `per-hop` format is:
 
 ~~~~
-+-------------+-------------+-------------+-----+------------+
-| n_1 payload | n_2 payload | n_3 payload | ... | hop filler |
-+-------------+-------------+-------------+-----+------------+
++----------------------+--------------------------+-------------------------------+--------------------+
+| channel_id (8 bytes) | amt_to_forward (4 bytes) | outgoing_cltv_value (4 bytes) | padding (16 bytes) |
```

Comment:
> Including the channel_id here is a bit of an abstraction leak. Why should the payment routing layer dictate exactly which channel should be utilized by the next node when forwarding a payment?

To me, the reason why the payment routing layer should use channels is that it actually only deals with channels. My opinion is that using `channel_id` is the natural thing to do, and using node identifier would be breaking the separation of layers.

> If [so], we allow the sender to dictate which channel rather than which node to forward the HTLC to, then we rob intermediate nodes of HTLC forwarding flexibility.

We don't, the forwarding node still can use whatever channel it prefers, right?

> From another angle: ideally, the routing layer in implementations treats multiple same-realm links between two nodes as a single link, and performs path finding taking this simplification into account.

But it doesn't seem very consistent with channel-centric failure messages (BOLT 4). Why do we return `0x1000` (UPDATE) errors if we only care about node-node links? What if the sender asks a route based on channel X and received an error/update for channel Y?

>Treating these as a single link simplifies path finding, and also allows intermediate forwarding nodes to freely fragment payments at the link level or, in a simpler case, have complete control as to which channels they wish to use to satisfy the forwarding requirements.

I have never understood how payment fragmentation works. How can the forwarding node split the onion?

> Also why 4 bytes for the amt_to_forward? This isn't enough to express satoshis, lets alone milli satoshis.

FWIW, this is consistent with the 4 bytes used in the `update_add_htlc` message. I think @rustyrussell wanted to put a hard upper limit on amounts.